pxr/imaging/hgiPresent/metal.mm

```diff
@@ -0,0 +1,318 @@
+// additional work to replace not that many lines of API specific code.
+
+namespace {
+PXR_NAMESPACE_USING_DIRECTIVE
+
+std::optional<MTLPixelFormat>
+_FindPreferredSurfaceFormatMatch(HgiFormat format, bool needs_sRGB_Transform)
+{
+    if (needs_sRGB_Transform) {
+        return MTLPixelFormatBGRA8Unorm_sRGB;
+    }
+
+    switch (format) {
+    case HgiFormatUNorm8Vec4:
+        return MTLPixelFormatBGRA8Unorm;
+    case HgiFormatUNorm8Vec4srgb:
+        return MTLPixelFormatBGRA8Unorm_sRGB;
+    case HgiFormatFloat16Vec4:
+        return MTLPixelFormatRGBA16Float;
+    case HgiFormatFloat32Vec3:
+    case HgiFormatFloat16Vec3:
+    case HgiFormatFloat32Vec4:
+        return MTLPixelFormatRGBA16Float;
+    case HgiFormatInvalid:
+    case HgiFormatCount:
+        TF_CODING_ERROR("Invalid Format");
+        return std::nullopt;
+    default:
+        return std::nullopt;
+    }
+}
+
+std::optional<CGColorSpaceRef>
+_MetalColorSpaceFromName(TfToken const &name)
+{
+    // Color spaces not in this map are unsupported.
+    // Nil means no color matching.
+    static const std::unordered_map<TfToken, CFStringRef, TfToken::HashFunctor>
+        namedColorSpaces = {
+            {GfColorSpaceNames->CIEXYZ, kCGColorSpaceGenericXYZ},
+            {GfColorSpaceNames->Data, nil},
+            {GfColorSpaceNames->Raw, nil},
+            {GfColorSpaceNames->Unknown, nil},
+            {GfColorSpaceNames->LinearAP1, kCGColorSpaceACESCGLinear},
+            {GfColorSpaceNames->LinearDisplayP3,
+                kCGColorSpaceExtendedLinearDisplayP3},
+            {GfColorSpaceNames->LinearRec2020,
+                kCGColorSpaceExtendedLinearITUR_2020},
+            {GfColorSpaceNames->LinearRec709, kCGColorSpaceLinearSRGB},
+            // See note about VK_COLOR_SPACE_BT709_NONLINEAR_EXT in vulkan.cpp
+            {GfColorSpaceNames->G18Rec709, kCGColorSpaceITUR_709},
+            {GfColorSpaceNames->G22AdobeRGB, kCGColorSpaceAdobeRGB1998},
+            // See note about VK_COLOR_SPACE_BT709_NONLINEAR_EXT in vulkan.cpp
+            {GfColorSpaceNames->G22Rec709, kCGColorSpaceITUR_709},
+            {GfColorSpaceNames->SRGBP3D65, kCGColorSpaceDisplayP3},
+            {GfColorSpaceNames->SRGBRec709, kCGColorSpaceSRGB}, // Extended?
+        };
+
+    if (const auto iter = namedColorSpaces.find(name);
+        iter != namedColorSpaces.end()) {
+        return iter-> second ? CGColorSpaceCreateWithName(iter->second) : nil;
+    }
+
+    return std::nullopt;
+}
+
+id<MTLComputePipelineState>
+_MakeComputePipeline(id<MTLDevice> device)
+{
+    NSString *shaderSource = @R"(
+#include <metal_stdlib>
+#include <simd/simd.h>
+
+using namespace metal;
+
+kernel void flipAndCopyTexture(
+    texture2d<float, access::read> inTexture,
+    texture2d<float, access::write> outTexture,
+    uint2 inCoord [[thread_position_in_grid]])
+{
+    if(inCoord.x >= outTexture.get_width() ||
+        inCoord.y >= outTexture.get_height()) {
+        return;
+    }
+
```

Comment:
was the intent to insert a matrix here to convert the color space to the presentation color space? GfColorSpace will give you the proper matrix, and it would be very inexpensive to insert it here. GfColorSpace will also give you parameters for the EOTF in the case of SRGB conversion from linear.

Reply:
No, this shader is needed for Metal because it doesn't support blitting to a different format or coordinate systems like Vulkan does. So we need to read the AOV to floats, flip vertically, then write back to the surface format.
But if we do decide that the presentation layer should do a best effort conversion to a supported window colour space, then yes it would go here.